libtsuba/src/RDGSlice.cpp

```diff
@@ -289,80 +425,79 @@ tsuba::RDGSlice::local_to_global_id() const {
 
 katana::Result<void>
 tsuba::RDGSlice::load_local_to_global_id() {
-  auto ltg_it = std::find_if(
-      core_->part_header().part_prop_info_list().begin(),
-      core_->part_header().part_prop_info_list().end(),
-      [](PropStorageInfo& psi) {
-        return psi.name() == RDGCore::kLocalToGlobalIDPropName;
-      });
-  if (ltg_it == core_->part_header().part_prop_info_list().end()) {
-    // some RDGs have no local to global array - use the default 0-length array
-    // instead
-    core_->set_local_to_global_id(katana::NullChunkedArray(arrow::uint64(), 0));
-    return katana::ResultSuccess();
-  }
-  std::vector<PropStorageInfo*> local_to_global{&(*ltg_it)};
-  KATANA_CHECKED_CONTEXT(
-      AddProperties(
-          core_->rdg_dir(), tsuba::NodeEdge::kNeitherNodeNorEdge, nullptr,
-          nullptr, local_to_global, nullptr,
-          [slice = this](const std::shared_ptr<arrow::Table>& props) {
-            return slice->core_->AddPartitionMetadataArray(props);
-          }),
-      "populating local to global id");
+  KATANA_CHECKED(demand_load_metadata_array(
+      RDGCore::kLocalToGlobalIDPropName,
+      [&]() {
+        core_->set_local_to_global_id(
+            katana::NullChunkedArray(arrow::uint64(), 0));
+      },
+      core_.get()));
+
   return katana::ResultSuccess();
 }
 katana::Result<void>
 tsuba::RDGSlice::load_local_to_user_id() {
-  auto ltu_it = std::find_if(
-      core_->part_header().part_prop_info_list().begin(),
-      core_->part_header().part_prop_info_list().end(),
-      [](PropStorageInfo& psi) {
-        return psi.name() == RDGCore::kLocalToUserIDPropName;
-      });
-  if (ltu_it == core_->part_header().part_prop_info_list().end()) {
-    // some RDGs have no local to user array - use the default 0-length array
-    // instead
-    core_->set_local_to_user_id(katana::NullChunkedArray(arrow::uint64(), 0));
-    return katana::ResultSuccess();
-  }
-  std::vector<PropStorageInfo*> local_to_user{&(*ltu_it)};
-  KATANA_CHECKED_CONTEXT(
-      AddProperties(
-          core_->rdg_dir(), tsuba::NodeEdge::kNeitherNodeNorEdge, nullptr,
-          nullptr, local_to_user, nullptr,
-          [slice = this](const std::shared_ptr<arrow::Table>& props) {
-            return slice->core_->AddPartitionMetadataArray(props);
-          }),
-      "populating local to global id");
+  KATANA_CHECKED(demand_load_metadata_array(
+      RDGCore::kLocalToUserIDPropName,
+      [&]() {
+        core_->set_local_to_user_id(
+            katana::NullChunkedArray(arrow::uint64(), 0));
+      },
```

Comment:
Oh, this is the empty array?  Seems like we are playing fast and loose with types, would arrow::null() also work?  If so, do we need to pass in a lambda?

Reply:
I think it is important that it be an empty array of `arrow::uint64()`.

Reply:
Ok, but weird to me to require a lambda that always returns the same thing.  Maybe it is only the same in the code for this PR.

Reply:
There are other metadata arrays that could make use of this function to load and unload on demand. But the other issue is that these lambdas both use the `core_->set_XXX()` interface. And there is no real connection between the string name of the local to global id array and the function `set_local_to_global_id()`. So we have to pass something to the function letting it know not only what the empty version of the array is, but how to set it.